File: scripts/used_by_students/PDE-scripts-2020/UnsteadyDiffusionExplicitVec.py
# Python

# Import necessary libraries
import numpy as np
import matplotlib.pyplot as plt
from scipy.sparse import diags

# Set parameters
Nx = 100
Nt = 40000
Dax = 1e-8
cL = 1
cR = 0
tend = 5000
length = 5e-3

# Calculate time and space steps
dt = tend/Nt
dx = length/Nx

# Calculate Fourier number
Fo = Dax*dt/(dx*dx)

# Create array of x values
x = np.linspace(0, length, Nx+1)

# Initialize matrix A
A = diags([-Fo, 1-2*Fo, -Fo], [-1,0,1], shape=(Nx+1, Nx+1))

# Initialize concentration array
c = np.zeros((Nt+1, Nx+1))

# Set boundary conditions
c[:, 0] = cL
c[:, Nx] = cR

# Transpose c for calculation convenience
c = c.T

# Perform time steps
for n in range(1, Nt+1):
    c[1:Nx, n] = A.dot(c[1:Nx, n-1])

# Transpose c back to original orientation
c = c.T

# Define desired output points
out = [12.6, 62.5, 125, 625, 5000]

# Calculate corresponding time steps
outdt = (np.array(out)//dt).astype(int)

# Plot results
plt.plot(x, c[outdt,:])
plt.xlabel('x')
plt.ylabel('Concentration')
plt.show()

File: scripts/PDE-scripts-2020/UnsteadyDiffusionExplicitVec.m
Nx = 100;
Nt = 40000;
Dax = 1e-8;
cL = 1;
cR = 0;
tend = 5000;
length = 5e-3;

dt = tend/Nt;
dx = length/Nx;

Fo = Dax*dt/(dx*dx);

x = linspace(0, length, Nx+1);

e = ones(Nx-1,1);
ld = [Fo*e; 0; 0];
md = [1; (1-2*Fo)*e; 1];
ud = [0; 0; Fo*e];
A = spdiags([ld md ud], [-1 0 1], Nx+1, Nx+1);

c = zeros(Nt+1,Nx+1);
c(:,1) = cL;
c(:,Nx+1) = cR;

c = c';
for n=1:Nt
    for i=2:Nx
       c(:,n+1) = A*c(:,n);
    end
end
c = c';

out = [12.6 62.5 125 625 5000];
outdt = fix(out/dt);
plot(x,c(outdt,:))